interop_binaries/build.rs

```diff
@@ -0,0 +1,62 @@
+fn main() {
+    // We only build the container image if the `testcontainer` feature is enabled, in order to
+    // avoid infinite recursion in our build process (since building the container image builds this
+    // package, among other things.)
+
+    #[cfg(feature = "testcontainer")]
+    {
+        use std::{env, process::Command};
+
+        // Build & save off a container image for the interop_aggregator.
+        // Note: `docker build` has an `--output` flag which writes the output to somewhere, which
+        // may be a tarfile. But `docker build --output` only exports the image filesystem, and not
+        // any other image metadata (such as exposed ports, the entrypoint, etc), so we can't easily
+        // use it.
+        let build_output = Command::new("docker")
+            .args([
+                "build",
+                "--file=Dockerfile.interop_aggregator",
+                "--quiet",
+                ".",
+            ])
+            .current_dir("..")
+            .env("DOCKER_BUILDKIT", "1")
+            .output()
+            .expect("Failed to execute `docker build` for interop aggregator");
+        assert!(
+            build_output.status.success(),
+            "Docker build of interop aggregator failed:\n{}",
+            String::from_utf8_lossy(&build_output.stderr)
+        );
+        let image_id = String::from_utf8(build_output.stdout).unwrap();
+        let image_id = image_id.trim();
+
+        let save_output = Command::new("docker")
+            .args([
+                "save",
+                &format!(
+                    "--output={}/interop_aggregator.tar",
+                    env::var("OUT_DIR").unwrap()
+                ),
+                image_id,
+            ])
+            .output()
+            .expect("Failed to execute `docker save` for interop aggregator");
+        assert!(
+            save_output.status.success(),
+            "Docker save of interop_aggregator failed:\n{}",
+            String::from_utf8_lossy(&build_output.stderr)
+        );
+
+        // Make a best-effort attempt to clean up after ourselves.
+        let rmi_output = Command::new("docker")
+            .args(["rmi", image_id])
+            .output()
+            .expect("Failed to execute `docker rmi` for interop aggregator");
+        assert!(
+            rmi_output.status.success(),
+            "Docker image removal of interop aggregator failed:\n{}",
+            String::from_utf8_lossy(&build_output.stderr)
+        );
```

Comment:
This is causing problems for me in rust-analyzer. I see the following error.

```
error: failed to run custom build command for `interop_binaries v0.1.0 (/home/david/Code/ppm/janus/interop_binaries)`

Caused by:
  process didn't exit successfully: `/home/david/Code/ppm/janus/target/debug/build/interop_binaries-b4d5fa1e6926891a/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'Docker image removal of interop aggregator failed:
  ', interop_binaries/build.rs:56:9
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
     1: core::panicking::panic_fmt
               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
     2: build_script_build::main
               at ./build.rs:56:9
     3: core::ops::function::FnOnce::call_once
               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I wonder if rust-analyzer is running the build script multiple times concurrently, and the later one stomps on the same image ID (since all the layers could be reused) and fails to delete it?  Perhaps we should give it a name and a randomly generated tag, and then delete by that tag.

Reply:
Interesting -- I didn't see anything like that. FWIW, if I change the build script to attempt deletion _twice_, I do see a similar crash on the second attempt, so I also suspect that a concurrency issue is at play.

Since image-removal is best-effort, I think the best solution is just to ignore failures. If the image has already been deleted, great! (Indeed, I think the common case of running `cargo test` will build the image/save it off/delete it during build, then load it again during tests, so the image still ends up in the local docker.)

However, I do wonder if this could run into trouble with concurrent builds in the same system: the `docker rmi` in one build could happen before the `docker save` in another build, causing the latter build to fail. But I think "parallel builds of the same code on the same machine" isn't something we need to support well. I am curious if you're still running into trouble with this change in place, however.

Reply:
I commented out the assertion on the `docker rmi` status and that got things working again.